pkg/core/image.go

```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return nil, err
+	}
+
+	imageMapper, err := newImageMapper(options.Registry, options.RegistryUser, imageManifest.Images)
+	if err != nil {
+		return nil, err
+	}
+
+	return imageMapper, nil
+}
+
+func relocateYAMLFile(yamlFile string, mapper *imageMapper, cwd string, outputPath string) (string, error) {
+	y, err := readYAMLFile(yamlFile, cwd)
+	if err != nil {
+		return "", err
+	}
+
+	output := outputFile(outputPath, yamlFile)
+	return filepath.Base(output), ioutil.WriteFile(output, mapper.mapImages(y), outputFilePermissions)
+}
+
+func readYAMLFile(yamlFile string, cwd string) ([]byte, error) {
+	u, err := url.Parse(yamlFile)
+	if err != nil {
+		return nil, err
+	}
+	if u.IsAbs() {
+		if u.Scheme == "file" {
+			return ioutil.ReadFile(u.Path)
+		}
+		// FIXME: Error: Get https:///storage.googleapis.com/knative-releases/serving/previous/v20180828-7c20145/istio.yaml: http: no Host in request URL
+		return downloadFile(u.String())
+	}
+
+	if !filepath.IsAbs(yamlFile) {
+		yamlFile = filepath.Join(cwd, yamlFile)
+	}
+	return ioutil.ReadFile(yamlFile)
+}
+
+func downloadFile(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func relocateManifest(manifestPath string, mapper *imageMapper, outputPath string) error {
+	if !isDirectory(outputPath) {
+		return fmt.Errorf("Invalid or non-existent output directory '%s'", outputPath)
+	}
+
+	manifest, err := NewManifest(manifestPath)
+	if err != nil {
+		return err
+	}
+
+	manifestDir := filepath.Dir(manifestPath)
+
+	outputManifest := &Manifest{
+		ManifestVersion: manifest.ManifestVersion,
+	}
+
+	outputManifest.Istio, err = relocateYamls(manifest.Istio, manifestDir, mapper, outputPath)
+	if err != nil {
+		return err
+	}
+	outputManifest.Knative, err = relocateYamls(manifest.Knative, manifestDir, mapper, outputPath)
+	if err != nil {
+		return err
+	}
+	outputManifest.Namespace, err = relocateYamls(manifest.Namespace, manifestDir, mapper, outputPath)
+	if err != nil {
+		return err
+	}
+
+	outputManifestPath := filepath.Join(outputPath, "manifest.yaml")
+	outputManifestBytes, err := yaml.Marshal(&outputManifest)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(outputManifestPath, outputManifestBytes, outputFilePermissions)
+}
+
+func relocateYamls(yamls []string, manifestDir string, mapper *imageMapper, outputPath string) ([]string, error) {
+	rel := []string{}
+	for _, y := range yamls {
+		// FIXME so this works for both relative paths and URLs
+		//err := relocateYAMLFile(filepath.Join(manifestDir, y), mapper, outputPath)
+		yRel, err := relocateYAMLFile(y, mapper, manifestDir, outputPath)
+		if err != nil {
+			return []string{}, err
+		}
+		rel = append(rel, yRel)
+	}
+	return rel, nil
+}
+
+func outputFile(outputPath string, yamlFile string) string {
```

Comment:
This silently wreaks havoc if two original URLs have the same basename

Reply:
Yes, I see what you mean. Not sure what we could do better though. Any suggestions?

Reply:
add a incrementing suffix? We control the reference to that file in the manifest

Reply:
Also, I wonder if we should enforce that the target dir either doesn't exist or is completely empty. Because if we add the behaviour I describe above, then repeated invocations will lead to more and more suffices being used. While only the relevant files will be linked in the general manifest, someone just zipping the whole directory will embark junk

Reply:
>add a incrementing suffix? We control the reference to that file in the manifest

`outputFile` is only called when `--yaml` (or whatever we end up calling it) is specified, so the manifest is kind of irrelevant. Also, I'd like the behaviour to be idempotent, as you suggested earlier. Also, if we output a non-obvious filename, we'll have to provide feedback to the user and if they don't pay attention (e.g. because they are running this command from a script), they could end up using the pre-existing output file. I think the behaviour is much more predictable, if a little dangerous in edge cases, if we stick with the basename rule.

> Also, I wonder if we should enforce that the target dir either doesn't exist or is completely empty. Because if we add the behaviour I describe above, then repeated invocations will lead to more and more suffices being used. While only the relevant files will be linked in the general manifest, someone just zipping the whole directory will embark junk[.]

I considered at the outset requiring an empty directory, but decided it's better for the command to be idempotent (in both `--yaml` and `--manifest` cases). If there's "junk" in the directory, the user may have put it there for good reason.

Reply:
> outputFile is only called when --yaml

I did not realize that, but even if true, I suspect the same kind of problem exists with the manifest case. What I meant (first of the problem and then of the solution) is this:

```yaml
istio:
- https://some.host/istio/awesome-crds.yaml
...
knative:
- https://knative.rocks.big.time/v1.2.3/awesome-crds.yaml
```

will end up with
```yaml
istio:
- awesome-crds.yaml
...
knative:
- awesome-crds.yaml
...
```

and a single `awesome-crds.yaml` file in the output directory (with contents depending of the order). Note that the problem also appears if there are homonymous files in the same category.


And so my "solution" is to generate something like
```yaml
istio:
- awesome-crds.yaml
...
knative:
- awesome-crds_1.yaml
...
```

Reply:
I agree there's a genuine problem with the `--manifest` case. However, I'd like to preserve idempotency and I don't really want to force the user to empty the directory before they repeat the command. One potential usecase is to add the manifest and relocated configuration files to a directory which already contains the image manifest and the binary images.

One solution is to append a hash, such as MD5, of the full path rather than an incrementing number. Of course, there's a risk of two _hashes_ clashing, but that's very unlikely. `ko` takes that approach when flattening image names. As a rather extreme safety net, we could detect when two hashes clash and fail the command.

Reply:
@zteve pointed out, over coffee, that collision detection should operate on input file names rather than taking into account output files. That way, idempotency can be preserved.

(He also spotted a nice edge case where an output k8s configuration file name clashes with the name of the manifest. Probably worth using subdirectories to avoid this altogether.)

Reply:
@ericbottard I pushed a new commit which prevents output file collisions when relocating manifests. It starts out with the base name of the file paths/URIs and, if those collide, appends a MD5 hash. If there's still a clash ([!](https://stackoverflow.com/questions/201705/how-many-random-elements-before-md5-produces-collisions)), it appends a SHA256 hash instead. If there's still a clash ([!!](https://stackoverflow.com/questions/4014090/is-it-safe-to-ignore-the-possibility-of-sha-collisions-in-practice)), the command fails with an error message.